Launching Spark over YARN

dennybritz opened the pull request:
**Usage Example:** 
SPARK_JAR=./core/target/spark-core-assembly-0.6.0-SNAPSHOT.jar ./run spark.deploy.yarn.Client **--jar** examples/target/scala-2.9.1/spark-examples_2.9.1-0.6.0-SNAPSHOT.jar **--class** spark.examples.SparkPi **--args** standalone **--num-workers** 3 **--worker-memory** 512M

**Notes:**
- Need to run sbt-assembly to generate the core-assembly jar used above.
- YARN does not support requesting containers/resources by the number of cores. Using `Runtime.getRuntime.availableProcessors()` for each slave for now. 
- The user must use a the String "standalone" as the master's URL. That starts the scheduler without trying to connect to a cluster.
- There are several TODOs. In particular 1. Checking for and handling container/app failure, 2. Letting the user set the context priority, and some code refactoring.
- Tested locally, but not on a real cluster.

tdas committed:
Added testcase for ByteBufferInputStream bugs.

Commit:
Various fixes when running on Mesos

tdas committed:
Fixed issues duplicate class issues in sbt assembly.

mateiz committed:
Upgraded to Akka 2 and fixed test execution (which was still parallel
across projects).

mateiz committed:
Ignore some IDE files

rrmckinley committed:
Scalacheck groupId has changed https://github.com/rickynils/scalacheck/issues/24. Necessary to build with scalaVersion 2.9.2. Works with 2.9.1 too.

mateiz committed:
Allow binding to a free port and change Akka logging to use SLF4J. Also
fixes various bugs in the previous code when running on Mesos.

mateiz committed:
Fixes to CoarseMesosScheduler

mateiz committed:
Increase the default wait time for EC2 clusters to 2 minutes.

mateiz committed:
Various code style fixes, mostly from IntelliJ IDEA

mateiz committed:
Initial framework to get a master and web UI up.

mateiz committed:
More work on deploy code (adding Worker class)

mateiz committed:
Detect connection and disconnection of slaves

mateiz committed:
More work on standalone deploy system.

mateiz committed:
Further updates to run processes on cluster.

mateiz committed:
Some refactoring to make cluster scheduler pluggable.

mateiz committed:
Miscellaneous bug fixes

mateiz committed:
Further refactoring, and start of a standalone scheduler backend

mateiz committed:
More work on standalone scheduler

mateiz committed:
Added back coarse-grained Mesos scheduler based on StandaloneScheduler.

mosharaf committed:
Removed Chained and Dfs broadcast implementations

mosharaf committed:
Compiles without Dfs

mateiz committed:
More work to allow Spark to run on the standalone deploy cluster.

mosharaf committed:
Removed some unused stuff

mosharaf committed:
Tweaks

mosharaf committed:
Refactored TrackMultipleValues out.

mosharaf committed:
Refactoring

mosharaf committed:
Refactoring

mosharaf committed:
Refactoring

mosharaf committed:
Slightly better check for isLocal

mosharaf committed:
Added stop() to the Broadcast subsystem

mateiz committed:
Added more methods for loading/saving with new Hadoop API

mateiz committed:
Formatting

Commit:
add Accumulatable, add corresponding docs & tests for accumulators

Commit:
improve scaladoc

Commit:
Accumulator now inherits from Accumulable, whcih simplifies a bunch of other things (eg., no +:=)

Conflicts:

	core/src/main/scala/spark/Accumulators.scala

Commit:
move Vector class into core and spark.util package

Commit:
add some functionality to Vector, delete copy in AccumulatorSuite

mateiz committed:
Fixes to coarse-grained Mesos scheduler in dealing with failed nodes

mateiz committed:
Allow null keys in Spark's reduce and group by

mosharaf committed:
Cleaned TreeBroadcast

mosharaf committed:
Cleaning BitTorrentBroadcast code...

mosharaf committed:
Merged with Upstream dev

mosharaf committed:
Fix for isLocal

Commit:
rename addToAccum to addAccumulator

Commit:
fix up scaladoc, naming of type parameters

Commit:
one more minor cleanup to scaladoc

dennybritz committed:
Always destroy SparkContext in after block for the unit tests.

Conflicts:

	core/src/test/scala/spark/ShuffleSuite.scala

JoshRosen committed:
Add Java API

Add distinct() method to RDD.

Fix bug in DoubleRDDFunctions.

mateiz committed:
Fix a bug where an input path was added to a Hadoop job configuration twice

JoshRosen committed:
Improve Java API examples

- Replace JavaLR example with JavaHdfsLR example.
- Use anonymous classes in JavaWordCount; add options.
- Remove @Override annotations.

JoshRosen committed:
Use mapValues() in JavaPairRDD.cogroupResultToJava().

JoshRosen committed:
Add type annotations to Java API methods.

Add missing Scala Map to java.util.Map conversions.

dennybritz committed:
Stlystic changes

Conflicts:

	core/src/test/scala/spark/MesosSchedulerSuite.scala

JoshRosen committed:
Minor cleanup and optimizations in Java API.

- Add override keywords.
- Cache RDDs and counts in TC example.
- Clean up JavaRDDLike's abstract methods.

JoshRosen committed:
Remove StringOps.split() from Java WordCount.

JoshRosen committed:
Detect non-zero exit status from PipedRDD process.

JoshRosen committed:
Add persist(), splits(), glom(), and mapPartitions() to Java API.

Commit:
tasks cannot access value of accumulator

mateiz committed:
Merge pull request #153 from JoshRosen/new-java-api

Java API

tdas committed:
Modified the block dropping code for better performance.

tdas committed:
Fixed bugs in block dropping code of MemoryStore and changed synchronized HashMap to ConcurrentHashMap in BlockManager.

tdas committed:
Fixed bug in BoundedMemoryCache.

tdas committed:
Fixed more stuff in BoundedMemoryCache.

tdas committed:
Added BlockRDD and a first-cut version of checkpoint() to RDD class.

tdas committed:
Merge branch 'dev' of github.com:tdas/spark into dev

mateiz committed:
Fixed Java union methods having same erasure.

Changed union() methods on lists to take a separate "first element"
argument in order to differentiate them to the compiler, because Java 7
considered it an error to have them all take Lists parameterized with
different types.

mosharaf committed:
Merge remote-tracking branch 'upstream/dev' into dev

Conflicts:
	core/src/main/scala/spark/broadcast/Broadcast.scala

mosharaf committed:
Broadcasts using BlockManager instead of BoundedMemoryCache

tdas committed:
Updated the new checkpoint RDD to remember partitioning of the original RDD.

mateiz committed:
Added a second version of union() without varargs.

mateiz committed:
Renamed test more descriptively

mateiz committed:
Add a sleep to prevent a failing test.

The BlockManager's put seems to be slightly asynchronous, which can
cause it to fail this test by not removing stuff from the cache before
we put the next value. We should probably change the semantics of put()
in this case but it's hard right now. It will also be hard for
asynchronously replicated puts.

mateiz committed:
Fixed SPARK_MEM not being passed when runner is java

mateiz committed:
Fixed AccumulatorSuite to clean up SparkContext with BeforeAndAfter

mosharaf committed:
Merge remote-tracking branch 'upstream/dev' into dev

paulcavallaro committed:
Logging Throwables in Info and Debug

Logging Throwables in logInfo and logDebug instead of swallowing them.

Conflicts:

	core/src/main/scala/spark/Logging.scala

mateiz committed:
Various reorganization and formatting fixes

mateiz committed:
Made BlockManagerMaster no longer be a singleton.

Also cleaned up a few formatting things throughout block manager code.

mateiz committed:
Merge branch 'dev' of git://github.com/tdas/spark into dev

mateiz committed:
Merge branch 'block-manager-fix' into dev

JoshRosen committed:
Fix Akka configuration in assembly jar.

This resolves an issue where running Spark from
the assembly jar would cause a "No configuration
setting found for key 'akka.version'" exception.

This solution is from the Akka Team Blog:

http://letitcrash.com/post/21025950392/

mosharaf committed:
Merge remote-tracking branch 'upstream/dev' into dev

mateiz committed:
Merge pull request #158 from JoshRosen/fix/assembly-akka-config

Merge Akka reference.conf files in sbt assembly task

mosharaf committed:
Merge remote-tracking branch 'upstream/dev' into dev

dennybritz committed:
Spark WebUI Implementation.

dennybritz committed:
Compact job page.

dennybritz committed:
Spark standalone mode cluster scripts.
Heavily inspired by Hadoop cluster scripts ;-)

dennybritz committed:
merge dev branch

dennybritz committed:
Webui improvements.

dennybritz committed:
Moved resources into static folder

JoshRosen committed:
Use sbt mergeStrategy for reference.conf files.

Cleans up #158 / 509b721.

mateiz committed:
Merge pull request #161 from JoshRosen/fix/assembly-akka-config

Use sbt mergeStrategy for reference.conf files.

dennybritz committed:
merge dev

shivaram committed:
Use maxMemory to better estimate memory available for BlockManager cache

shivaram committed:
Avoid an extra partition copy by passing an iterator to blockManager.put

mateiz committed:
Merge pull request #162 from shivaram/dev

Use maxMemory to better estimate memory available for BlockManager cache

dennybritz committed:
Mostly stlyistic changes.

dennybritz committed:
Use spray's implicit Marshaller for Futures.

mateiz committed:
Merge branch 'dev' of github.com:mesos/spark into dev

Conflicts:
	project/SparkBuild.scala

dennybritz committed:
Merged standalone and mesos EC2 scripts

dennybritz committed:
Settings variables and bugfix for stop script.

dennybritz committed:
merged dev branch

dennybritz committed:
Read config variables before to get the master port

harveyfeng committed:
Fix for partitioning when sorting in descending order

mateiz committed:
Added a unit test for cross-partition balancing in sort, and changes to
RangePartitioner to make it pass. It turns out that the first partition
was always kind of small due to how we picked partition boundaries.

mateiz committed:
Made range partition balance tests more aggressive.

This is because we pull out such a large sample (10x the number of
partitions) that we should expect pretty good balance. The tests are
also deterministic so there's no worry about them failing irreproducibly.

shivaram committed:
Fix testcheckpoint to reuse spark context defined in the class

mateiz committed:
Merge pull request #165 from shivaram/dev

Fix test checkpoint to reuse spark context defined in the class

mosharaf committed:
Bug fixed. Broadcast now works with BlockManager.

mosharaf committed:
Merge remote-tracking branch 'upstream/dev' into dev

mosharaf committed:
Merge branch 'dev' into bc-bm

dennybritz committed:
working on yarn

dennybritz committed:
Add Apache license to non-trivial scripts taken from Hadoop.

mateiz committed:
Merge pull request #160 from dennybritz/clusterscripts

Standalone cluster scripts

mosharaf committed:
Merge remote-tracking branch 'upstream/dev' into dev

mosharaf committed:
BroadcastManager is a class now (replaced Braodcast object)

shivaram committed:
Avoid a copy in ShuffleMapTask by creating an iterator that will be used by the
block manager.

mateiz committed:
Merge pull request #166 from shivaram/dev

Avoid a copy in ShuffleMapTask

shivaram committed:
Use JavaConversion to get a scala iterator

shivaram committed:
Use an implicit conversion to get the scala iterator

dennybritz committed:
Working on yarn

dennybritz committed:
Merge dev

mateiz committed:
Removed separate launcher for EC2 standalone cluster

JoshRosen committed:
Print exit status in PipedRDD failure exception.

mateiz committed:
Merge pull request #167 from JoshRosen/piped-rdd-fixes

Detect non-zero exit status from PipedRDD process

mateiz committed:
Merge pull request #168 from shivaram/dev

Use JavaConversion to get a scala iterator

shivaram committed:
Changes to make size estimator more accurate. Fixes object size, pointer size
according to architecture and also aligns objects and arrays when computing
instance sizes. Verified using Eclipse Memory Analysis Tool (MAT)

shivaram committed:
Add link to Java wiki which specifies what changes with compressed oops

shivaram committed:
Use HotSpotDiagnosticMXBean to get if CompressedOops are in use or not

shivaram committed:
If spark.test.useCompressedOops is set, use that to infer compressed oops
setting. This is useful to get a deterministic test case

shivaram committed:
Update test cases to match the new size estimates. Uses 64-bit and compressed
oops setting to get deterministic results

mateiz committed:
Changed multi-get method in BlockManager to return an iterator

mateiz committed:
Switch to Akka futures in connection manager.

It's still not good because each Future ends up waiting on a lock, but
it seems to work better than Scala Actors, and more importantly it
allows us to use onComplete and other listeners on futures.

mateiz committed:
Return remotely fetched blocks in a pipelined fashion from BlockManager

mateiz committed:
Use Promises instead of having a Future wait on a thread in
ConnectionManager.

mateiz committed:
Renamed apply() to call() in Java API and allowed it to throw Exceptions

shivaram committed:
Move object size and pointer size initialization into a function to enable unit-testing

shivaram committed:
Add test-cases for 32-bit and no-compressed oops scenarios.

dennybritz committed:
working on yarn

dennybritz committed:
Merge branch 'dev' of git://github.com/mesos/spark into yarn

dennybritz committed:
Some code cleanups

dennybritz committed:
Some code cleanups. Ignore log4j properties file

dennybritz committed:
remove properties file

dennybritz committed:
Pass SPARK_JAVA_OPTS as extra argument

shivaram committed:
Change BlockManagerSuite test cases to use a deterministic size estimator and
update the results to match the new estimates

dennybritz committed:
Pass extra java options taken from SPARK_JAVA_OPTS

dennybritz committed:
Explicitly set JVM memory on slaves

mateiz committed:
Merge pull request #171 from shivaram/for-size-estimator-pull

Size estimator changes for dev

dennybritz committed:
rsync root directory in EC2 script

dennybritz committed:
Use root login in standalone AMI

mateiz committed:
Merge pull request #172 from dennybritz/dev

Rsync root directory in EC2 script

mateiz committed:
Added other SparkContext constructors to JavaSparkContext

mateiz committed:
Merge pull request #147 from mosharaf/dev

Broadcast refactoring/cleaning up

mateiz commented on core/src/main/scala/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -0,0 +1,197 @@
+      || System.getProperty("spark.master.port").toInt == 0) {
+      Thread.sleep(10)
+    }
+    logInfo("Port: " + System.getProperty("spark.master.port"))
```
This might not always work well, because the user may have set spark.master.port from outside (e.g. as a JVM option), so you go ahead and launch workers, but those workers might connect _before_ the Spark master is started. Ideally you would keep looking for that property and try to open a connection to that port once it's set, and wait for the master to really be up. How hard would it be to do that?

dennybritz replied:
Good point, I'll fix that.

mateiz commented on core/src/main/scala/spark/deploy/yarn/WorkerRunnable.scala:
```diff
@@ -0,0 +1,120 @@
+    return env
+  }
+  
+  def conntectToCM : ContainerManager = {
```
This is misspelled

dennybritz replied:
Good catch.

mateiz commented on core/src/main/scala/spark/deploy/yarn/ApplicationMaster.scala:
```diff
@@ -0,0 +1,197 @@
+          val workerHostname = container.getNodeId().getHost()
+          // YARN does not support requesting resources by the number of cores yet.
+          // TODO: How do we handle this?
+          // val workerCores = 1
```
Good question. I would go for asking the number of cores per worker as a command-line argument instead of setting it to Runtime.getNumberOfProcessors as you do now, just because the latter might be extremely high. Shouldn't be too hard to change.

dennybritz replied:
Okay, I'll add that. However, if the user resource request cannot be satisfied (there are not enough free cores) we won't be able to tell the user in advance.

mateiz replied:
Yeah, this is fine, but I just want to give users a way to limit the number of cores. For example, with the old code, many users might fit into the memory of one machine, and they would all try to use the max number of cores.

On Aug 23, 2012, at 9:08 PM, Denny wrote:

> In core/src/main/scala/spark/deploy/yarn/ApplicationMaster.scala:
> 
> > -      val workersToRequest = math.max(args.numWorkers - numWorkersRunning.intValue, 0)
> > -      val amResp = allocateWorkerResources(workersToRequest).getAMResponse()
> > -      val allocatedContainers = amResp.getAllocatedContainers
> > -      if (allocatedContainers.size > 0) {
> > -       logInfo("Allocated " + allocatedContainers.size + " containers.")
> > -       logInfo("Cluster Resources: " + amResp.getAvailableResources)
> > -        // Run each of the allocated containers
> > -        for (container <- allocatedContainers) {
> > -          val masterUrl = "akka://spark@%s:%s/user/%s".format(
> > -            System.getProperty("spark.master.host"), System.getProperty("spark.master.port"),
> > -            StandaloneSchedulerBackend.ACTOR_NAME)
> > -          val workerId = numWorkersRunning.intValue.toString
> > -          val workerHostname = container.getNodeId().getHost()
> > -          // YARN does not support requesting resources by the number of cores yet.
> > -          // TODO: How do we handle this?
> > -          // val workerCores = 1
> >   Okay, I'll add that. However, if the user resource request cannot be satisfied (there are not enough free cores) we won't be able to tell the user in advance.
> 
> —
> Reply to this email directly or view it on GitHub.

mateiz committed:
Fixed test failures due to broadcast not stopping correctly

mateiz committed:
Fix further issues with tests and broadcast.

The broadcast fix is to store values as MEMORY_ONLY_DESER instead of
MEMORY_ONLY, which will save substantial time on serialization.

mateiz committed:
Set log level in tests to WARN

JoshRosen committed:
Cache points in SparkLR example.

mateiz committed:
Merge pull request #179 from JoshRosen/fix/sparklr-caching

Cache points in SparkLR example

dennybritz committed:
Merge branch 'dev' into yarn

dennybritz committed:
Fix typo in conntectToCM.

dennybritz committed:
Added workerCores as argument.

dennybritz committed:
Removed cores comment.

mateiz commented:
Hey Denny,

Let me know when you've made the changes above and I'll merge this. It would be nice to have this in the main repo.

dennybritz committed:
Waiting for the spark master to be reachable.

dennybritz committed:
Adding worker-cores argument to client.

dennybritz commented:
I made the changes. One thing that I couldn't quite figure out is why the application is marked as "FAILED" by YARN even though the ApplicationMaster exits with exit code 0. I am even setting the status to FinalApplicationStatus.SUCCEEDED explicitly. The error message isn't very enlightening: `Application application_1346087295431_0014 failed 1 times due to AM Container for appattempt_1346087295431_0014_000001 exited with  exitCode: 0 due to: .Failing this attempt.. Failing the application.` Any ideas?

mateiz commented:
Thanks. Let's look into that later.
